Feature/product

SwanyCastle opened the pull request:
### Pull Request

> ### 제목
>
> - Feat: 상품 등록, 수정, 삭제 기능 구현
> - Fix: 멤버 정보 조회시 Self-Invocation 문제 발생 수정
> - Fix: 회원정보 조회, 수정, 삭제시 본인 확인
> - Feat: 전체 상품 목록 조회 및 특정 판매자 상품 목록 조회 기능 구현

> ### 변경사항
> 
> 📌 **AS-IS**
> - Entity 이름 변경 및 로그인 기능추가
> - 회원 조회, 수정, 삭제 기능 구현 및 테스트코드 작성
> - 로그인시 토큰 Redis에서 관리 하도록 수정, 토큰이 필요한 요청을 받을시 Redis 에 해당 토큰이 존재하는지 토큰 유효성 검사 추가
> - 로그아웃 기능 구현
> - 코드 스타일에 맞게 코드 정렬
> - OAuth 2.0 를 활용하여 소셜 로그인 구현
>
> 🔖 **TO-BE**
>
> 🔧 상품 등록, 수정, 삭제 기능 구현
> - 상품 등록, 수정, 삭제 기능에 대한 api endpoint 접근을 SELLER 권한을 가진 사용자에게만 허용
>   하기위해 PreAuthorize 어노테이션 사용 해당 어노테이션 사용위해 SecurityConfig.java 에 EnableMethodSecurity 어노테이션추가
> - 상품 등록, 수정, 삭제 기능 트랜잭션 처리되도록 구현
> - 상품 등록, 수정, 삭제 기능에 필요한 각 DTO 클래스 작성 및 응답 코드 추가
> - 상품 등록, 수정, 삭제 기능처리시 커스텀 에러처리를 위한 ProductException 추가
>
> 🔧 멤버 정보 조회시 Self-Invocation 문제 발생 수정
> - Transactional 어노테이션이 없는 getMemberDetails 메서드가
>    Transactional 어노테이션이 있는 getMemberByMemberId 메서드를 호출하게되면
>    getMemberByMemberId 는 트랜잭션 처리가 되지 않는 문제가 발생해서
>    getMemberDetails 에도 Transactional 어노테이션을 붙여 수정
>
> 🔧 회원 정보 조회, 수정, 삭제 시 요청으로 들어온 Path Variable 의 memberId 와  헤더값의 토큰 에있는 memberId 값을 비교
> - 회원정보 조회, 수정, 삭제등 요청으로 들어온 헤더값의 토큰을 추출해
>   토큰에 있는 Member Id 와 Path Variable 로 들어온 Member Id 가 같은지 확인하는 로직을
>   AuthService.java 에 정의해서 AuthService 메서드로 추출해 다른 서비스에서 사용할 수 있도록 변경
> - 그에 따른 테스트 코드도 같이 변경
>
> 🔧 전체 상품 목록 조회 및 특정 판매자 상품 목록 조회 기능 구현
> - 전체 상품 목록 조회
>   Pageable 을 사용해 페이지네이션 구현 (테스트를 위해 PAGE_SIZE 를 5로 지정)
>   page, search, status, ordering 을 쿼리 파라미터 required = false 로 받아
>   default 값을 설정해 두고 들어오는 값만을 가지고 상품 목록 조회
> - 특정 판매자 상품 목록 조회
>   Pageable 을 사용해 페이지네이션 구현 (테스트를 위해 PAGE_SIZE 를 5로 지정)
>   memberId 를 패스 파라미터로 받고
>   page, search, status, ordering 을 쿼리 파라미터 required = false 로 받아
>   default 값을 설정해 두고 들어오는 값만을 가지고 memberId 를 가지는 member 의 상품 목록 조회
>
> - ProductDto.Response 에서 MemberDto 의 member 로 멤버 정보 보여주던 것을
>   seller 변수에 memberId 만 표시되도록 변경
>
> - UpdateProductDto 에서 stockQuantity, price 최소값 0으로 설정 (음수 받지 못하게)
> 
> - 상품 등록, 수정, 삭제시 회원 조회, 수정, 삭제와 마찬가지로
>   요청으로 들어온 헤더에서 토큰 값을 추출해 토큰 안에있는 memberId 와 패스파라미터로 받은 memberId
>   비교해 동일 사용자의 요청인지 확인
>

> ### 테스트
>
> - [ ] 테스트 코드
> - [x] API 테스트

SwanyCastle committed:
[Feat]: 상품 등록, 수정, 삭제 기능 구현

상품 등록, 수정, 삭제 기능 구현
- 상품 등록, 수정, 삭제 기능에 대한 api endpoint 접근을 SELLER 권한을 가진 사용자에게만 허용
  하기위해 PreAuthorize 어노테이션 사용 해당 어노테이션 사용위해 SecurityConfig.java 에 EnableMethodSecurity 어노테이션추가
- 상품 등록, 수정, 삭제 기능 트랜잭션 처리되도록 구현
- 상품 등록, 수정, 삭제 기능에 필요한 각 DTO 클래스 작성 및 응답 코드 추가
- 상품 등록, 수정, 삭제 기능처리시 커스텀 에러처리를 위한 ProductException 추가

SwanyCastle committed:
[Fix]: 멤버 정보 조회시 Self-Invocation 문제 발생 수정

@Transactional 어노테이션이 없는 getMemberDetails 메서드가
@Transactional 어노테이션이 있는 getMemberByMemberId 메서드를 호출하게되면
getMemberByMemberId 는 트랜잭션 처리가 되지 않는 문제가 발생해서
getMemberDetails 에도 @Transactional 어노테이션을 붙여 수정

SwanyCastle committed:
[Fix]: 회원정보 조회, 수정, 삭제시 본인 확인

회원정보 조회, 수정, 삭제시 본인 확인
- 회원정보 조회, 수정, 삭제등 요청으로 들어온 헤더값의 토큰을 추출해
  토큰에 있는 Member Id 와 Path Variable 로 들어온 Member Id 가 같은지 확인하는 로직을
  AuthService.java 에 정의해서 AuthService 메서드로 추출해 다른 서비스에서 사용할 수 있도록 변경
- 그에 따른 테스트 코드도 같이 변경

SwanyCastle committed:
[Feat]: 전체 상품 목록 조회 및 특정 판매자 상품 목록 조회 기능 구현

전체 상품 목록 조회 및 특정 판매자 상품 목록 조회 기능 구현
- 전체 상품 목록 조회
  Pageable 을 사용해 페이지네이션 구현 (테스트를 위해 PAGE_SIZE 를 5로 지정)
  page, search, status, ordering 을 쿼리 파라미터 required = false 로 받아
  default 값을 설정해 두고 들어오는 값만을 가지고 상품 목록 조회
- 특정 판매자 상품 목록 조회
  Pageable 을 사용해 페이지네이션 구현 (테스트를 위해 PAGE_SIZE 를 5로 지정)
  memberId 를 패스 파라미터로 받고
  page, search, status, ordering 을 쿼리 파라미터 required = false 로 받아
  default 값을 설정해 두고 들어오는 값만을 가지고 memberId 를 가지는 member 의 상품 목록 조회

ProductDto.Response 에서 MemberDto 의 member 로 멤버 정보 보여주던 것을
seller 변수에 memberId 만 표시되도록 변경

UpdateProductDto 에서 stockQuantity, price 최소값 0으로 설정 (음수 받지 못하게)

상품 등록, 수정, 삭제시 회원 조회, 수정, 삭제와 마찬가지로
요청으로 들어온 헤더에서 토큰 값을 추출해 토큰 안에있는 memberId 와 패스파라미터로 받은 memberId
비교해 동일 사용자의 요청인지 확인

hjkim22 commented:
내부 호출에서도 트랜잭션을 사용해 데이터 일관성을 유지하는군요.
트랜잭션이 필요하지 않다고 생각했던 메서드에도 필요한 이유를 알게 되었습니다!
코드 잘봤습니다!

SwanyCastle committed:
[Fix]: 상품 등록, 수정시 전달받는 RequestBody 에 Valid 어노테이션 추가

SwanyCastle committed:
[Fix]: 상품 등록 및 ProductDto.Request 에서 상품 상태 값 제거

상품 등록 수정 및 ProductDto.Request 에서 상품 상태 값 제거
- DB 에 상품 가격 저장시 BigDecimal.valueOf() 사용해서 double 형태로 저장 하고 status 값을 ProductStatus.IN_STOCK 으로 초기화
- 상품을 등록할때 ProductStatus.IN_STOCK 으로 초기화 할 예정이라 굳이 ProductDto.Request 에서 받을 필요가 없어서 제거

spring-kotlin-lover commented on src/main/java/com/ecommerce/controller/ProductController.java:
```diff
@@ -0,0 +1,96 @@
+  public List<ProductDto.Response> getProductList(
+      @RequestParam(required = false, defaultValue = "1") Integer page,
+      @RequestParam(required = false, defaultValue = "") String search,
+      @RequestParam(required = false, defaultValue = "") String status,
```
Status 는 Enum class 를 그대로 선언하여 사용해보는 것은 어떨까요?
ordering 또한 마찬가지 입니다. 

Enum class 를 그대로 지정하여 사용하면 데이터 변환되는 횟수를 줄일 수 있고, string 보다는 데이터셋이 정해져잇으니 관리하기에도 용이할 것 같습니다.

spring-kotlin-lover commented on src/main/java/com/ecommerce/service/product/ProductServiceImplement.java:
```diff
@@ -0,0 +1,292 @@
+   */
+  @Override
+  @Transactional(readOnly = true)
+  public List<ProductDto.Response> getProductListByMemberId(
```
특별한 이유가 없다면 List 보다는 Page 객체를 반환하여 client 가 response 에서도 인지할 수 있게 해주세요~!

spring-kotlin-lover commented on src/main/java/com/ecommerce/service/product/ProductServiceImplement.java:
```diff
@@ -0,0 +1,292 @@
+
+    authService.equalToMemberIdFromToken(product.getMember().getMemberId(), token);
+
+    if (updateRequest.getProductName() != null && !updateRequest.getProductName().isEmpty()) {
+      product.setProductName(updateRequest.getProductName());
+    }
+
+    if (updateRequest.getDescription() != null && !updateRequest.getDescription().isEmpty()) {
+      product.setDescription(updateRequest.getDescription());
+    }
+
+    if (updateRequest.getStockQuantity() != null) {
+      product.setStockQuantity(updateRequest.getStockQuantity());
+    }
+
+    if (updateRequest.getPrice() != null) {
+      product.setPrice(updateRequest.getPrice());
+    }
+
+    if (updateRequest.getStatus() != null) {
+      product.setStatus(updateRequest.getStatus());
+    } else {
+      if (product.getStockQuantity() == 0) {
+        product.setStatus(ProductStatus.NO_STOCK);
+      } else {
+        product.setStatus(ProductStatus.IN_STOCK);
+      }
+    }
```
Nullable 여부를 체크하면서 optional 하게 update 하는 것보다는 Non-null 로 지정받고, 요청한대로 update 하면 어떨까요? 특정 필드를 선택하여 업데이트해달라는 요청인데, client 입장에서도 어떤 필드가 바뀌었는지 체크해서 요청해야되니 번거롭고, server 입장에서도 어떤 필드가 바뀌는 것으로 요청오는 것인지 체크해야되니 더 번거로운 로직이 들어가는 것 같아요.

if 문이 많아진다면 이것을 줄일수 있는 방법에 대해 고민해보거나, 로직을 심플하게 가져가는 방안을 고려해보면 좋을 것 같아요.

SwanyCastle committed:
[Fix]: PR 리뷰 코멘트에 대한 부분 수정

- RequestParam 으로 받는 status, ordering 을 String 말고 Enum 클래스 자체로 받게 수정
- ordering 은 변수이름을 sortType 으로 변경하고 SortType Enum 클래스를 생성하여 관리
- 페이지네이션 한 결과 값을 반환 할 때 List<ProductDto.Response> -> Page<ProductDto.Response> 로 변경
- 상품 정보 수정시에 Path 로 api 요청을 받아 Optional 하게 상품 정보를 변경하던 부분을 Put 으로
  요청받아 UpdateProductDto 의 멤버변수들을 전체 Non-null 하게 지정받아 수정하도록 변경
